adds null safety codelab (#2983)
dart-lang/site-www: src/codelabs/null-safety.md

---
title: Null safety codelab
description: Learn about and practice writing null safe code in DartPad!
js: [{url: 'https://dartpad.dev/inject_embed.dart.js', defer: true}]
---
{% assign useIframe = false -%}
<?code-excerpt replace="/ *\/\/\s+ignore_for_file:[^\n]+\n//g; /(^|\n)*\/\/\s+ignore:[^\n]+\n/$1/g; /(\n[^\n]+) *\/\/\s+ignore:[^\n]+\n/$1\n/g"?>
<?code-excerpt plaster="none"?>

> 

Reviewer: Throughout this codelab, please cross-link any concept (like "late" or "definite assignment") to the reference docs section for that concept in https://dart.dev/null-safety/understanding-null-safety

> This codelab teaches you about Dart’s null-safe type system,

Reviewer: s/Dart's null-safety type system/Dart's sound null safety feature/

(The content here is beyond the type system, and also talks about the tooling that supports the type system)

> which was introduced in Dart SDK 2.12. When you opt into null safety,

Reviewer: Please s/Dart SDK 2.12/Dart 2.12/

types in your code are non-nullable by default,
meaning that values can’t be `null` unless you say they can be.

This codelab covers the following material:

* Nullable and non-nullable types.
* How and when to use the null-aware operators: `?` and `!`
* How the `late` keyword affects variables and initialization.

Using embedded DartPad editors, you can test your knowledge by
running example code and completing exercises. To get the most out of
this codelab, you should have some knowledge of [basic Dart syntax](/samples).

{{site.alert.note}}
  This page uses embedded DartPads to display examples and exercises.
  {% include dartpads-embedded-troubleshooting.md %}
{{site.alert.end}}

## What is null safety?

> Dart's null-safe type system makes types in your code non-nullable by default.

Reviewer: s/Dart's null-safe type system makes types in your code non-nullable by default./With sound null safety, types are non-nullable by default/

That means values in your Dart programs can't be `null` unless you say they can be.
With null safety, you can find null reference errors before you even
run your program.

> Here are some core principles of sound null safety in the Dart language:

Reviewer: These's aren't really the "core principles" of null safety; we have those here:
https://dart.dev/null-safety#null-safety-principles

Can we just say something a bit more general like "The core traits of sound null safety in Dart are:" ?

-   All types are non-nullable by default.
-   Values can't be null unless you say they can be.
-   Finding null errors at edit time is much better than finding them at run
    time.

## Nullable and non-nullable types

> When you opt into null safety, all types are non-nullable by default.

Reviewer: Link "opt into" to https://dart.dev/null-safety#enable-null-safety

For example, if you have a variable of
type `String`, it will always contain a string.

If you want a variable of type `String` to accept any string

> or the value `null`, but nothing else,

Reviewer: ", but nothing" else seems superfluous here.

  List<String> aListofNullableStrings = ['one', null, 'three'];

  print('aListofStrings is $aListofStrings.');
  print('aNullableListOfStrings is $aNullableListOfStrings.');
  print('aListofNullableStrings is $aListofNullableStrings.');
}
```

## Using ?, !, and late

Null safety introduces new operators and keywords for dealing with null values,
`?`, `!`, and `late`. Here's how you use them:

-   If you're creating a variable that can hold a null value, you can use a
    `?` to inform Dart of the variable’s nullability.
-   If you're sure that an expression with a nullable type isn't null, you can 
    add a postfix exclamation mark (`!`) to make Dart treat it as non-nullable.
-   If you want to initalize a non-nullable variable to a
    non-null value before it’s used, insert `late` before the
    variable’s type.

### Example: The assertion operator (!)

If you'd like to assign a nullable expression to a variable that's
non-nullable, you can use the assertion operator (the exclamation point: `!`).
By adding `!` just after the expression, you tell Dart that the value won't be
null, and that it's safe to assign it to a non-nullable variable. If you're

> wrong, Dart throws an exception.

Reviewer: s/an exception/an exception at run-time/ to make it very clear.

Reviewer: Maybe also add: "This makes the `!` operator unsafe, and you should only use it when you are very sure that the expression isn't null."

  final Coach coach;
}

class Coach {
  final Team team;
}

void main() {
  final myTeam = Team();
  final myCoach = Coach();
  myTeam.coach = myCoach;
  myCoach.team = myTeam;

  print('All done!');
}
```

### Exercise: Late and lazy

Here's another pattern that `late` can help with: lazy initialization for
expensive non-nullable fields. Try running this code without changing it. What
do you think will change if you make `_cache` a `late` field?

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/late_lazy.dart"?>
```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
int _computeValue() {
  print('Computing value...');
  return 3;
}

class CachedValueProvider {
  final _cache = _computeValue();
  int get value => _cache;
}

void main() {
  print('Calling constructor...');
  var provider = CachedValueProvider();
  print('Getting value...');
  print('The value is ${provider.value}!');
}
```

Fun fact: After you add `late` to the declaration of `_cache`, you can move the
`_computeValue` function into the `CachedValueProvider` class and the code will
still work! Initialization expressions for `late` fields can use instance
methods in their initializers.

> ## Type promotion

Reviewer: I think this should be above the Late section; promotion is a more basic concept that should be understood first.

> With null safety, Dart takes null checks into account. Nullable variables that

Reviewer: I'd rewrite this first sentence to something like "With sound null safety, the [flow analysis](https://dart.dev/null-safety/understanding-null-safety#flow-analysis) has been extended to take nullability into account.

}

void main() {
  print(getLength('This is a string!'));
}
```

### Exercise: Promotion with exceptions

Promotion works with exceptions as well as return statements. Try a null check
that throws an `Exception` instead of returning zero.

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/promotion_exceptions.dart" replace="/.*if\ \(.*\n.*\n.*//g"?>
```dart:run-dartpad:ga_id-promotion_exceptions:null_safety-true
int getLength(String? str) {
  // Try throwing an exception here if `str` is null.

  return str.length;
}

void main() {
  print(getLength(null));
}
```

## The late keyword

Sometimes variables — fields in a class, or top-level variables — 
should be non-nullable, but they can't be
assigned a value immediately. For cases like that, use the
[`late` keyword]( /null-safety/understanding-null-safety#late-variables).

When you put `late` in front of a variable declaration,
that tells Dart the following:

-   Don't assign that variable a value yet.
-   You *will* assign it a value later.
-   You'll make sure that the variable has a value *before* the variable is used.

If you declare a variable `late` and the variable is read before it's assigned a

> value, a `LateInitializationError` is thrown.

Reviewer 2: Unfortunately this is no longer a guarantee. This error has been removed from the SDK and the only behavior that is defined is that "some" error will be thrown. Refer to https://github.com/dart-lang/sdk/issues/44361#issuecomment-737672074 for more information.

```suggestion
value, an error is thrown by Dart.
```

Reviewer 3: I removed "by Dart" because if you're going to use the passive voice, you should go passive all the way. :) (We should either say "an error is thrown" or "Dart throws an error", which isn't much more precise.)

---
title: Null safety codelab
description: Learn about and practice writing null safe code in DartPad!
js: [{url: 'https://dartpad.dev/inject_embed.dart.js', defer: true}]
---
{% assign useIframe = false -%}
<?code-excerpt replace="/ *\/\/\s+ignore_for_file:[^\n]+\n//g; /(^|\n)*\/\/\s+ignore:[^\n]+\n/$1/g; /(\n[^\n]+) *\/\/\s+ignore:[^\n]+\n/$1\n/g"?>
<?code-excerpt plaster="none"?>
<style>
{% comment %}
TODO(chalin): move this into one of our SCSS files
{% endcomment -%}
iframe[src^="https://dartpad"] {
  border: 1px solid #ccc;
  margin-bottom: 1rem;
  min-height: 150px;
  resize: vertical;
  width: 100%;

> }

Author: it seems like this conflicts with this: https://github.com/dart-lang/site-www/blob/master/src/_assets/css/_code.scss#L21-L28

i'm going to remove the styling from this page and see how it looks wit the one in `css/_code.scss`.

  a = null;
  print('a is $a.');
}
```

### Example: Nullable types

What if you need a variable that *can* hold a null value?  Try changing the
type of `a` so that `a` can be either null or an int:

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/nullable_types.dart" replace="/int\?/int/g"?>
```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
void main() {
  int a;
  a = null;
  print('a is $a.');
}
```

### Example: More nullable types!

Type parameters for generics can also be nullable or non-nullable. Try using
question marks to correct the type declarations of `aNullableListOfStrings` and
`aListOfNullableStrings`:

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/more_nullable_types.dart" replace="/String\?/String/g"?>
```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
void main() {
  List<String> aListofStrings = ['one', 'two', 'three'];
  List<String> aNullableListOfStrings = [];
  List<String> aListofNullableStrings = ['one', null, 'three'];

  print('aListofStrings is $aListofStrings.');
  print('aNullableListOfStrings is $aNullableListOfStrings.');
  print('aListofNullableStrings is $aListofNullableStrings.');
}
```

> ## Using ? and !

Reviewer 3: We've already introduced `?`, and we don't want to encourage incorrect usage of either `!` or `late`, so how about we move this section to be just before `late` (like it is in null-safe DartPad)?

```suggestion
## Using the null assertion operator (!)
```

Reviewer 3: @ryanmcginnis I don't see this comment addressed in your changes.

> Null safety introduces new operators and keywords for dealing with null values,
> `?`, `!`, and `late`. Here's how you use them:
> 
> -   If you're creating a variable that can hold a null value, use a
>     question mark (`?`) to inform Dart of the variable’s nullability.

Reviewer 3: ```suggestion
```

-   If you're sure that an expression with a nullable type isn't null, use a
    postfix exclamation mark (`!`, or "bang operator") to
    make Dart treat it as non-nullable.
-   If you want to initalize a non-nullable variable to a
    non-null value before it’s used, use the `late` keyboard before the
    variable’s type.

### Example: The assertion operator (!)

If you'd like to assign a nullable expression to a variable that's
non-nullable, you can use the
[null assertion operator](/null-safety/understanding-null-safety#null-assertion-operator)
(the exclamation point: `!`). By adding `!` just after the expression,
you tell Dart that the value won't be null, and that it's safe to assign it to a
non-nullable variable.

But if you're wrong, Dart throws an exception at run-time. This makes the
`!` operator unsafe, and you should only use it when you are very sure that the
expression isn't null.

In the example below, try adding exclamation points to correct the

> three broken assignments:

Reviewer 3: I only see two broken assignments! Which is surprising, since I suspect that `a` had been broken, but apparently the analyzer can tell that `couldBeNullButIsnt` has a non-null value.

```suggestion
broken assignments:
```

Reviewer 3: Nor this one

been extended to take nullability into account. Nullable variables that
can't possibly contain null values are treated like non-nullable variables.
This behavior is called
[type promotion](/null-safety/understanding-null-safety#type-promotion-on-null-checks).

### Example: Definite assignment

Dart's type system can track where variables are assigned and
where their values are read, and to verify that non-nullable fields are given
values before any code tries to read from them. This process is called
[definite assignment](/null-safety/understanding-null-safety#definite-assignment-analysis).

Try uncommenting the if-else statement in the code below, and
watch the analyzer errors disappear:

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/definite_assignment.dart" replace="/if/\/\/if/g; /\ \ text\ =/\/\/  text =/g; /\ \ \}/  \/\/}/g"?>
```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
void main() {
  String text;

  //if (DateTime.now().hour < 12) {
  //  text = "It's morning! Let's make aloo paratha!";
  //} else {
  //  text = "It's afternoon! Let's make biryani!";
  //}

  print(text);
  print(text.length);
}
```

### Exercise: Null checking

> In the example below, add an if statement to the beginning of `getLength` that

Reviewer 3: Or this one

returns zero if `str` is null:

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/type_promotion.dart" replace="/.*if\ \(.*\n.*\n.*//g"?>
```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
int getLength(String? str) {
  // Add null check here

  return str.length;
}

void main() {
  print(getLength('This is a string!'));
}
```

### Exercise: Promotion with exceptions

Promotion works with exceptions as well as return statements. Try a null check
that throws an `Exception` instead of returning zero.

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/promotion_exceptions.dart" replace="/.*if\ \(.*\n.*\n.*//g"?>
```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
int getLength(String? str) {
  // Try throwing an exception here if `str` is null.

  return str.length;
}

void main() {
  print(getLength(null));
}
```

## Late keyword

> Sometimes fields in a class *should* be non-nullable, but those fields can't be

Reviewer 3: ```suggestion
Sometimes variables — fields in a class, or top-level variables — 
should be non-nullable, but they can't be
```

assigned a value immediately. For cases like that, use the
[`late` keyword]( /null-safety/understanding-null-safety#late-variables).

`late` is how you tell Dart that:

-   You aren't going to assign that field a value right away.
-   But you *are* going to assign it a value later.
-   And you'll make sure it's assigned a value *before* it's accessed.

If you declare a field `late` and the field is read before it is assigned a
value, a `LateInitializationError` is thrown to tell you what went wrong.

### Example: The `late` keyword

Try using the `late` keyword to correct the following code. For a little extra
fun afterward, try commenting out the line that sets `description`!

> 

Reviewer 3: Have you noticed that when you click **Reset**, analysis isn't run again? I've filed https://github.com/dart-lang/dart-pad/issues/1785 about this.

}
```

### Exercise: Late and lazy

Here's another pattern that `late` can help with:
[lazy initialization](/null-safety/understanding-null-safety#lazy-initialization)
for expensive non-nullable fields. Try running this code without changing it.
What do you think will change if you make `_cache` a `late` field?

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/late_lazy.dart"?>
```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
int _computeValue() {
  print('Computing value...');
  return 3;
}

class CachedValueProvider {
  final _cache = _computeValue();
  int get value => _cache;
}

void main() {
  print('Calling constructor...');
  var provider = CachedValueProvider();
  print('Getting value...');
  print('The value is ${provider.value}!');
}
```

Fun fact: After you add `late` to the declaration of `_cache`, you can move the
`_computeValue` function into the `CachedValueProvider` class and the code will
still work! Initialization expressions for `late` fields can use instance
methods in their initializers.

## What's next?

Congratulations, you've finished the codelab! If you'd like to learn more, here
are some suggestions for where to go next:

> -   [Play with DartPad](https://nullsafety.dartpad.dev/).

Reviewer 3: ```suggestion
-   [Play with DartPad]({{site.dartpad}}).
```

> -   Try another [codelab](https://dart.dev/codelabs).

Reviewer 3: ```suggestion
-   Try another [codelab](/codelabs).
```

-   Learn more about null safety:

>     -   Overview: [Sound null safety](https://dart.dev/null-safety).

Reviewer 3: ```suggestion
    -   Overview: [Sound null safety](/null-safety).
```

    -   Deep dive:

>         [Understanding null safety](https://dart.dev/null-safety/understanding-null-safety).

Reviewer 3: ```suggestion
        [Understanding null safety](/null-safety/understanding-null-safety).
```

> -   [Get the Dart SDK](https://dart.dev/get-dart).

Reviewer 3: ```suggestion
-   [Get the Dart SDK](/get-dart).
```

If you're interested in using embedded DartPads, like this codelab does, see

> [best practices for using DartPad in tutorials](https://dart.dev/resources/dartpad-best-practices).

Reviewer 3: ```suggestion
[best practices for using DartPad in tutorials](/resources/dartpad-best-practices).
```

---

> title: "Null safety"

Reviewer 3: Let's have a different name from the main null safety page (also, we don't need quotation marks here):

```suggestion
title: Null safety codelab
```

description: Learn about and practice writing null safe code in DartPad!
js: [{url: 'https://dartpad.dev/inject_embed.dart.js', defer: true}]
---
{% assign useIframe = false -%}
<?code-excerpt replace="/ *\/\/\s+ignore_for_file:[^\n]+\n//g; /(^|\n) *\/\/\s+ignore:[^\n]+\n/$1/g; /(\n[^\n]+) *\/\/\s+ignore:[^\n]+\n/$1\n/g"?>
<?code-excerpt plaster="none"?>
<style>
{% comment %}
TODO(chalin): move this into one of our SCSS files
{% endcomment -%}
iframe[src^="https://dartpad"] {
  border: 1px solid #ccc;
  margin-bottom: 1rem;
  min-height: 150px;
  resize: vertical;
  width: 100%;
}
</style>

> This codelab teaches you about Dart’s null safe type system, which was introduced in Dart SDK 2.12. When you opt into null safety, types in your code are non-nullable by default, meaning that values can’t be null unless you say they can be.

Reviewer 3: Before I review the rest of this, could you please do a pass over it to use 80-char-max lines with [semantic line breaks](https://github.com/dart-lang/site-shared/blob/master/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks)?

Author: done! PTAL

Using embedded DartPad editors, you can test your knowledge by running example code and completing exercises. To get the most out of this codelab, you should have some knowledge of [basic Dart syntax](/samples).

This codelab covers the following material:

* Nullable and non-nullable types.

> * How and when to use the null-aware operators: ? and !

Reviewer 4: ```suggestion
* How and when to use the null-aware operators: `?` and `!`
```

Reviewer 4: Also, are they really called null-aware operators? Just checking. I was always using the term for `??` and `?.`.

> * How the late keyword affects variables and initialization.

Reviewer 4: ```suggestion
* How the `late` keyword affects variables and initialization.
```

{{site.alert.note}}
  This page uses embedded DartPads to display examples and exercises.
  {% include dartpads-embedded-troubleshooting.md %}
{{site.alert.end}}

## What is null safety?

Dart's null safe type system makes types in your code non-nullable by default,

> meaning that values can't be null unless you say they can be. Null safety helps

Reviewer 4: [nit] I'd add one more sentence, saying the same thing again, some other way. These sentences are super hard to parse.

Author: done!

prevent null reference errors. 

Here are some core principles
of sound null safety in the Dart language:

-   All types are non-nullable by default.
-   Values can't be null unless you say they can be.
-   Finding null errors at edit time is much better than finding them at run
    time.

## Non-nullable types

All types are non-nullable by default. For example, if you have a variable of
type String, it will always contain a string.

## Nullable types

> If you want a variable of type String to accept any string, or the value null

Reviewer 4: ```suggestion
If you want a variable of type `String` to accept any string, or the value null
```

I'll stop suggesting, but, in general, I tend to add backticks around every type and piece of code in any given paragraph.

  List<String> aListofStrings = ['one', 'two', 'three'];
  List<String> aNullableListOfStrings = [];
  List<String> aListofNullableStrings = ['one', null, 'three'];

  print('aListofStrings is $aListofStrings.');
  print('aNullableListOfStrings is $aNullableListOfStrings.');
  print('aListofNullableStrings is $aListofNullableStrings.');
}
```

### Example: Definite assignment

Dart's type system is crafty enough to track where variables are assigned and
where their values are read, and to verify that non-nullable fields are given
values before any code tries to read from them. This process is called
"definite assignment."

Try uncommenting the if-else statement in the code below, and
watch the analyzer errors disappear:

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/definite_assignment.dart" replace="/if/\/\/if/g; /\ \ text\ =/\/\/  text =/g; /\ \ \}/  \/\/}/g"?>
```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
void main() {
  String text;

  //if (DateTime.now().hour < 12) {
  //  text = "It's morning! Let's make aloo paratha!";
  //} else {
  //  text = "It's afternoon! Let's make biryani!";
  //}

  print(text);
  print(text.length);
}
```

> ## New null-aware operators: ? and !

Author: sounds good to me. we seem to call them operators elsewhere, though. for example: https://github.com/dart-lang/site-www/pull/2983

"New operators and keywords related to null safety include `?`, `!`, and `late`."

should we update that topic?

> ## New null-aware operators: ? and !

Author: maybe we could say "syntax"? i'm not sure.

> Null safety introduces two operators, ? and !. Here's how you use them:

Reviewer 3: Rewrite this intro... somehow... Maybe something about dealing with the possibility of null values? (So this isn't about new operators, it's about syntax you're likely to see that we haven't already discussed.)

Author: done, PTAL

> -   If you're creating a variable that can have the value null, add a
>     question mark (?) to the variable's type declaration. Types are
>     non-nullable by default and made nullable by adding ?.

Reviewer 3: This seems like it's rehashing the nullable types stuff from above.

```suggestion
```

-   If you're accessing a property and the operand might be null, use the

>     null-aware version member access operator (?.).

Author: i meant to say "null-aware version of the conditional member access operator." i saw it described as such in this topic: https://dart.dev/null-safety#using-variables-and-expressions

"Once you opt into null safety, you can’t use the member access operator (.) if the operand might be null. Instead, you can use the null-aware version of that operator (?.)"

is that correct?

-   If you're sure that an expression with a nullable type isn't null, you can 
    add a postfix exclamation mark (!) to make Dart treat it as non-nullable.

### Example: Conditional access

Conditional access is a handy way to tighten up code that needs to read
properties that could be null. For example, the following expression evaluates
to the value of `b` as long as `a` isn't null. If `a` is null, then the

> expression evaluates to null:

Author: i removed this section. but i should try that translate test more often. for example, "tighten up" definitely does not localize well!

  a?.b

Without null safety, *two* conditional access operators are needed for the
following code to work: one after `big` and one after `little`. With null
safety, conditional access can short-circuit, so this expression only requires a

> single `?`:

Reviewer 3: This example is a little hard to understand because it's so hypothetical. We've already added `?` to a type up above, and we don't show `?.`, so it's unclear what new thing this is meant to show.

Author: yeah i think it's best to remove the conditional access section.

Author: yeah i agree. i think i'll remove this section.

}

class LittleThing {
  int fetchInt() => 12;
}

void main() {
  BigThing? big = BigThing();

  print('The value is:');
  print(big.little.fetchInt());
}
```

### Example: The assertion operator `!`

If you'd like to assign a nullable expression to a variable that's
non-nullable, you can use the assertion operator (the exclamation point: `!`).
By adding `!` just after the expression, you tell Dart that the value won't be
null, and that it's safe to assign it to a non-nullable variable. If you're
wrong, Dart throws an exception.

In the example below, try adding exclamation points to correct the
three broken assignments:

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/assertion_operator.dart" replace="/first!/first/g; /!.abs/.abs/g"?>
```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
int? couldReturnNullButDoesnt() => -3;

void main() {
  int? couldBeNullButIsnt = 1;
  List<int?> listThatCouldHoldNulls = [2, null, 4];

  int a = couldBeNullButIsnt;
  int b = listThatCouldHoldNulls.first; // first item in the list
  int c = couldReturnNullButDoesnt().abs(); // absolute value

  print('a is $a.');
  print('b is $b.');
  print('c is $c.');
}
```

> 

Reviewer 3: This seems like a new section on type promotion. Should we add a section header move the first paragraph from the "Type promotion" exercise up here?

Author: done. LMK if that's what you had in mind.

---
title: Null safety codelab
description: Learn about and practice writing null safe code in DartPad!
js: [{url: 'https://dartpad.dev/inject_embed.dart.js', defer: true}]
---
{% assign useIframe = false -%}
<?code-excerpt replace="/ *\/\/\s+ignore_for_file:[^\n]+\n//g; /(^|\n)*\/\/\s+ignore:[^\n]+\n/$1/g; /(\n[^\n]+) *\/\/\s+ignore:[^\n]+\n/$1\n/g"?>
<?code-excerpt plaster="none"?>
<style>
{% comment %}
TODO(chalin): move this into one of our SCSS files
{% endcomment -%}
iframe[src^="https://dartpad"] {
  border: 1px solid #ccc;
  margin-bottom: 1rem;
  min-height: 150px;
  resize: vertical;
  width: 100%;
}
</style>

This codelab teaches you about Dart’s null safe type system,
which was introduced in Dart SDK 2.12. When you opt into null safety,
types in your code are non-nullable by default,
meaning that values can’t be `null` unless you say they can be.

Using embedded DartPad editors, you can test your knowledge by
running example code and completing exercises. To get the most out of

> this codelab, you should have some knowledge of [basic Dart syntax](/samples).

Reviewer 3: Move this paragraph below the bulleted list? (Or change the intro to lead in more naturally. I found the transition jarring.)

Author: done

This codelab covers the following material:

* Nullable and non-nullable types.
* How and when to use the null-aware operators: `?` and `!`
* How the `late` keyword affects variables and initialization.

{{site.alert.note}}
  This page uses embedded DartPads to display examples and exercises.
  {% include dartpads-embedded-troubleshooting.md %}
{{site.alert.end}}

## What is null safety?

> Dart's null safe type system makes types in your code non-nullable by default.

Reviewer 3: ```suggestion
Dart's null-safe type system makes types in your code non-nullable by default.
```

That means values in your Dart programs can't be `null` unless you say they can be.

> With null safety, you can fix `null` reference errors at edit-type, before you even

Reviewer 3: ```suggestion
With null safety, you can find null reference errors before you even
```

run your program.

Here are some core principles of sound null safety in the Dart language:

-   All types are non-nullable by default.
-   Values can't be null unless you say they can be.
-   Finding null errors at edit time is much better than finding them at run
    time.

> ## Non-nullable types

Reviewer 3: ```suggestion
## Nullable and non-nullable types
```

> All types are non-nullable by default. For example, if you have a variable of

Reviewer 3: ```suggestion
When you opt into null safety, all types are non-nullable by default.
For example, if you have a variable of
```

> type String, it will always contain a string.

Reviewer 3: ```suggestion
type `String`, it will always contain a string.
```

> ## Nullable types

Reviewer 3: ```suggestion
```

> If you want a variable of type `String` to accept any string, or the value null,
> but nothing else, you give the variable a nullable type by adding a `?` after the

Reviewer 3: Hard to read this diff, but let's delete one comma and delete one "you"

```suggestion
If you want a variable of type `String` to accept any string
or the value `null`, but nothing else,
give the variable a nullable type by adding a `?` after the
```

type name. For example, a variable of type `String?` can contain a string, or
it can be null.

### Example: Introducing non-nullable types

The variable `a` below is declared as an `int`. Try changing the value in the
assignment to 3 or 145. Anything but null!

<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/non_nullable_types.dart" replace="/145/null/g"?>

> ```dart

Reviewer 3: It looks like you need to add the dartpad code to these snippets. Something like this:

`dart:run-dartpad:ga_id-nonnullable_type:null_safety-true`

Author: done!
